sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -119,6 +119,15 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
 
   object HashAggregation extends Strategy {
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
+      // We can cancel the map aggregation due to GC problems
+      case logical.Aggregate(groupingExpressions, aggregateExpressions, child)
+        if !sqlContext.conf.partAggregationEnabled =>
+        execution.Aggregate(
```

Comment:
I think that the decision of whether to pre-aggregate or not is orthogonal to whether we should use a generated aggregate or the regular interpreted aggregate operator.  Therefore, if we decide to implement this feature then I think we'll have to refactor this block in order to support both operators.  Maybe there's a nice common interface / builder pattern that we could use to abstract over the actual aggregate operator that's used.